tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/TerraformPackage.java

```diff
@@ -95,45 +102,31 @@ public String execute(Properties properties) throws TerraformException {
       logger.debug(String.format("tfModulesPath is %1$s", tfModulesPath.toAbsolutePath().toString()));
 
+      if (targetTfRootPath.toFile().exists()) FileUtils.forceDelete(targetTfRootPath.toFile());
+      FileUtils.forceMkdir(targetTfRootPath.toFile());
+
       File tfSourceFile = Paths.get("src", "main", "tf").toFile();
+
       Path tfRootPath = !StringUtils.isEmpty(tfRootDir)
           ? TerraformUtils.getTerraformRootModuleDir(tfRootDir)
           : (tfSourceFile.exists() && tfSourceFile.isDirectory() && tfSourceFile.listFiles().length > 1
-              ? tfSourceFile.toPath()
-              : TerraformUtils.getDefaultTerraformRootModuleDir());
+          ? tfSourceFile.toPath()
+          : TerraformUtils.getDefaultTerraformRootModuleDir());
       logger.debug(String.format("tfRootPath is %1$s", tfRootPath.toAbsolutePath().toString()));
 
-      List<Path> tfVarFilePaths = StringUtils.isEmpty(tfVarFiles)
-          ? new ArrayList<>()
-          : Arrays.stream(tfVarFiles.split(","))
-          .map(x -> tfRootPath.resolve(x.trim()))
-          .collect(Collectors.toList());
-      logger.debug(String.format("tfVarFiles is %1$s", tfVarFilePaths));
+      copyTfRootDir(tfRootPath, targetTfRootPath);
 
-      //copy tfRoot directory to target
-      if (targetTfRootPath.toFile().exists()) FileUtils.forceDelete(targetTfRootPath.toFile());
+      copyTfVarFiles(tfRootPath, tfVarFiles, targetTfRootPath);
 
-      FileUtils.forceMkdir(targetTfRootPath.toFile());
-      List<Path> tfRootDirFiles = Files.walk(tfRootPath, 1)
-          .filter(path -> !path.equals(tfRootPath))
-          .collect(Collectors.toList());
-
-      for (Path file : tfRootDirFiles.stream().filter(f -> !excludedFiles.contains(f.getFileName().toString())).collect(Collectors.toList())) {
-        if (file.toFile().isDirectory()) {
-          FileUtils.copyDirectory(file.toFile(), targetTfRootPath.resolve(file.getFileName().toString()).toFile(), new FileFilter() {
-            @Override
-            public boolean accept(File pathname) {
-              return excludedFiles.stream().noneMatch(excludedFile -> pathname.getName().endsWith(excludedFile));
-            }
-          });
-        } else {
-          Files.copy(file, targetTfRootPath.resolve(file.getFileName().toString()));
-        }
-      }
+      saveTfVarsAsTfVarFile(tfRootPath, tfVars, targetTfRootPath, "properties.auto.tfvars.json");
 
-      for (Path tfVarFilePath : tfVarFilePaths) {
-        Files.copy(tfVarFilePath, targetTfRootPath.resolve(tfVarFilePath.getFileName().toString().replaceAll(".tfvars", "").concat(".auto.tfvars")));
-      }
+      Map<String,String> envTfVars = System.getenv()
+          .entrySet()
+          .stream()
+          .filter(map -> map.getKey().startsWith("TF_VAR_"))
+          .collect(Collectors.toMap(map -> map.getKey().substring(7), map -> map.getValue()));
+
+      saveTfVarsAsTfVarFile(tfRootPath, envTfVars, targetTfRootPath, "pipeline.auto.tfvars.json");
```

Comment:
I guess I didn't realize we were talking about baking this into the tf-maven-plugin. In this case, I don't think we should name the output file 'pipeline.auto.tfvars.json' since these values might not always come from a build pipeline or something called a pipeline. Maybe we name it something more descriptive like 'env.auto.tfvars.json' ?

Reply:
> TerraformPackage has just gotten massive because of all the additional functionality we have added into it - hardly open/closed. We need to refactor this guy. Decorator pattern comes to mind. Decorator would allow some or all of this functionality to be reused without changing the class - it decouples the functional bits presently layered on top of each other.

I was thinking all these like TerraformPackage should be like domain objects and use composition to factor out some of this code

Reply:
Yea im not big fan of putting DTCC stuff into the maven plugin. This seems like something that belongs in pipeline code